Kafka/Confluent with Airflow - Basic tutorial (#1510)
astronomer/docs: learn/airflow-kafka.md

:::

1. In `kafka_example_dag_1`, add the following import statement:

    ```python
    from airflow_provider_kafka.operators.await_message import AwaitKafkaMessageOperator
    ```

2. Copy and past the following code at the end of your DAG:

    ```python
    def await_function(message):
        if isinstance(json.loads(message.value()), int):
            if json.loads(message.value()) % 5 == 0:
                return f" Got the following message: {json.loads(message.value())}"

    await_message = AwaitKafkaMessageOperator(
        task_id=f"awaiting_message_in_{my_topic}",
        topics=[my_topic],
        # the apply function needs to be passed with its location for the triggerer
        apply_function="provider_example_code.await_function", 
        kafka_config={
            **connection_config
            "group.id": "awaiting_message",
            "enable.auto.commit": False,
            "auto.offset.reset": "beginning",
        },
        xcom_push_key="retrieved_message",
    )

>     consumer_task >> await_message

Reviewer: Shouldn't the `producer_task` created in Step 2 still be in this DAG?

            task_id=f"produce_to_{my_topic}",
            topic=my_topic,
            producer_function=producer_function, 
            kafka_config=connection_config
        )
    ```

    The code above retrieves the environment variables you defined in [Step 1](#step-1-configure-your-astro-project) and packages them into a configuration dictionary that can be used by the ProduceToTopicOperator. Any Python function which returns a generator can be passed to the `producer_function` parameter of the ProduceToTopicOperator. Make sure your producer function returns a generator that contains key-value pairs where the value is in a format your Kafka topic accepts as input. In this example, the generator produces a JSON value. Additionally, if you have defined a schema for your Kafka topic, the generator needs to return compatible objects.

3. Run your DAG.

4. View the logs of your task instance. The 5 produced events will be listed.

    ![Producer logs](/img/guides/kafka-producer-logs.png)

5. View the produced events in your Kafka cluster. The example screenshot below shows them in the Confluent Cloud.

    ![Producer logs](/img/guides/confluent-produced-tasks.png)

:::info

> If you are using a Kafka cluster that is running on your local machine, you will need to adjust the `connection_config` parameter in this example. To connect to the [Kafka quick start](https://kafka.apache.org/documentation/#quickstart) cluster when running Airflow in Docker, set the following properties in your Kafka cluster's `server.properties` file:

Reviewer: I found this part a little confusing since it's instructing to update the parameter directly in the `connection_config` in the DAG, but in the DAG code we're pulling the value from an env variable that we create in Step 1. Wondering if this note should be a step above, maybe as part of Step 1. Or below substep 2 in this step and be super clear that we're saying make the adjustment directly to the DAG file.

Author: Yeah you are right, I changed the config so that all parts that need adjustments are passed as env variables and put the note in Step 1, as well as the note about how to change server.properties in the Prerequisites, I feel this is a bit more logical, so that the prereq is "Kafka server running with correct configs" before step 1. I hope it is clearer now :)

> ```text

Reviewer: Not sure if this would be included in the Kafka knowledge prereq (which I definitely don't have), but I got tripped up for a minute because I added these lines to `server.properties` after my cluster was already running, and didn't realize I needed to restart the cluster for them to take effect. I'll leave it up to you whether we should add something like "before starting your cluster" to the end of the last sentence, or whether you think any Kafka users would already know this :)

Author: Very good point. I had forgotten to restart too at first 😅

- `topics`: A list of topics or regex patterns to read from.
- `apply_function`: A Python function that is applied to all messages that are read. If the function returns any data the task will be ended and marked as successful. The returned data will be pushed to XCom unless the BaseOperator argument `do_xcom_push` is set to `False`.
- `apply_function_args`: Positional arguments for the `apply_function`.
- `apply_function_kwargs`: Keyword arguments for the `apply_function`.
- `kafka_config`: The configuration for the Kafka client, including the connection information. For a full list of parameters please refer to the [librdkafka GitHub repository](https://github.com/edenhill/librdkafka/blob/master/CONFIGURATION.md).
- `poll_timeout`: The amount of time in seconds that the task should wait for a message in its active state.
- `poll_interval`: The amonut of time in seconds that the task should wait in the deferred state.
- `xcom_push_key`: The key under which to save the returned data to XCom.

## Best practices

> Apache Kafka is a tool optimized for streaming messages at high frequencies for example in an IoT application. Airflow is designed to handle orchestration of data pipelines in batches at a frequency of once per minute or less.

Reviewer: ```suggestion
Apache Kafka is a tool optimized for streaming messages at high frequencies, for example in an IoT application. Airflow is designed to handle orchestration of data pipelines in batches at a frequency of no more than once per minute.
```

---
title: "Use Kafka with Airflow"
sidebar_label: "Kafka/Confluent"
description: "How to produce to and consume from Kafka topics using the Airflow Kafka provider"
id: airflow-kafka
---

> import Tabs from '@theme/Tabs';
> import TabItem from '@theme/TabItem';
> 

Reviewer 2: ```suggestion
```

    librdkafka-dev
    ```

3. Add the following two packages to your `requirements.txt` file:

    ```text
    confluent-kafka==1.8.2
    airflow-provider-kafka
    ```

:::info

If you are running Airflow as a standalone application and using an M1 Mac please the modified installation instrustions in the README of the [`airflow-provider-kafka`](https://github.com/astronomer/airflow-provider-kafka) package.

:::

4. In the .env file define the following environment variables. Provide your own Kafka topic name and boostrap server. If you are connecting to a cloud based Kafka cluster you will likely also need to provide an API Key and API Secret:

    ```text
    KAFKA_TOPIC_NAME=my_kafka_topic_name
    BOOSTRAP_SERVER=my_boostrap_server
    KAFKA_API_KEY=my_api_key
    KAFKA_API_SECRET=my_API_secret
    ```

5. Run the following command to start your project in a local environment:

    ```sh
    astro dev start
    ```

## Step 2: Create a DAG with a producer task

The [Airflow Kafka provider package](https://github.com/astronomer/airflow-provider-kafka) contains an operator allowing you to produce events directly to an existing Kafka topic, the ProduceToTopicOperator.

> 1. Create a new Python file in your `dags` folder called `kafka_example_dag_1.py`.

Reviewer 2: ```suggestion
1. Create a new file in your `dags` folder called `kafka_example_dag_1.py`.
```

    from airflow_provider_kafka.operators.produce_to_topic import ProduceToTopicOperator

    # get the topic name from .env
    my_topic = os.environ["KAFKA_TOPIC_NAME"]

    # get Kafka configuration information
    connection_config = {
        "bootstrap.servers": os.environ["BOOSTRAP_SERVER"],
        "security.protocol": "SASL_SSL", # adjust for local clusters
        "sasl.mechanism": "PLAIN", #  adjust for local clusters
        "sasl.username": os.environ["KAFKA_API_KEY"], # adjust for local clusters
        "sasl.password": os.environ["KAFKA_API_SECRET"] # adjust for local clusters
    }

    with DAG(
        dag_id="kafka_example_dag_1",
        start_date=datetime(2022, 11, 1),
        schedule=None,
        catchup=False,
    ):

        # define the producer function
        def producer_function():
            for i in range(5):
                yield (json.dumps(i), json.dumps(i+1))

        # define the producer task
        producer_task = ProduceToTopicOperator(
            task_id=f"produce_to_{my_topic}",
            topic=my_topic,
            producer_function=producer_function, 
            kafka_config=connection_config
        )
    ```

>     The code above retrieves the environment variables you defined in [Step 1](#step-1-configure-your-astro-project) and packages them into a configuration dictionary that can be used by the ProduceToTopicOperator. Any Python function which returns a generator can be passed to the `producer_function` parameter of the ProduceToTopicOperator. Make sure your producer function returns a generator that contains key-value pairs where the value is in a format your Kafka topic accepts as input. In the example above the generator produces a JSON value. Additionally, if you have defined a schema for your Kafka topic, the generator needs to return compatible objects.

Reviewer 2: ```suggestion
    The code above retrieves the environment variables you defined in [Step 1](#step-1-configure-your-astro-project) and packages them into a configuration dictionary that can be used by the ProduceToTopicOperator. Any Python function which returns a generator can be passed to the `producer_function` parameter of the ProduceToTopicOperator. Make sure your producer function returns a generator that contains key-value pairs where the value is in a format your Kafka topic accepts as input. In this example, the generator produces a JSON value. Additionally, if you have defined a schema for your Kafka topic, the generator needs to return compatible objects.
```

>     If you are connecting to a local Kafka cluster you might need to adjust the `connection_config` dictionary (see tip below).

Reviewer 2: ```suggestion
```

3. Run your DAG.

4. View the logs of your task instance. The 5 produced events will be listed.

    ![Producer logs](/img/guides/kafka-producer-logs.png)

5. View the produced events in your Kafka cluster. The example screenshot below shows them in the Confluent Cloud.

    ![Producer logs](/img/guides/confluent-produced-tasks.png)

> :::tip

Reviewer 2: ```suggestion
:::info
```

> If you are using a Kafka cluster that is running locally on your machine you will need to adjust the `connection_config` parameter. To connect to the [Kafka quick start](https://kafka.apache.org/documentation/#quickstart) cluster when running Airflow in Docker, set the following properties in your Kafka cluster's `server.properties` file:

Reviewer 2: ```suggestion
If you are using a Kafka cluster that is running on your local machine, you will need to adjust the `connection_config` parameter in this example. To connect to the [Kafka quick start](https://kafka.apache.org/documentation/#quickstart) cluster when running Airflow in Docker, set the following properties in your Kafka cluster's `server.properties` file:
```

listeners=PLAINTEXT://:9092,DOCKER_HACK://:19092
advertised.listeners=PLAINTEXT://localhost:9092,DOCKER_HACK://host.docker.internal:19092

> listener.security.protocol.map=PLAINTEXT:PLAINTEXT,DOCKER_HACK:PLAINTEXT

Reviewer 2: Add ` ```text ` styling here

> while using `"bootstrap.servers":"host.docker.internal:19092"` and `"security.protocol": "PLAINTEXT"` as your `connection_config`. Learn more in [Confluent's documentation](https://www.confluent.io/blog/kafka-client-cannot-connect-to-broker-on-aws-on-docker-etc/#scenario-5).

Reviewer 2: ```suggestion
Then add `"bootstrap.servers":"host.docker.internal:19092"` and `"security.protocol": "PLAINTEXT"` to your `connection_config` parameter in the DAG.
```

:::

## Step 3: Add a consumer task

> With the ConsumeFromTopicOperator Airflow is able to consume messages from topics. 

Reviewer 2: ```suggestion
The ConsumeFromTopicOperator enables Airflow to consume messages from topics. 
```

> 1. Add the following import statement to the `kafka_example_dag_1`.

Reviewer 2: ```suggestion
1. Add the following import statement to `kafka_example_dag_1`.
```

    ```python
    from airflow_provider_kafka.operators.consume_from_topic import ConsumeFromTopicOperator
    ```

> 2. In the `kafka_example_dag_1` add the following code below the `producer_task`.

Reviewer 2: ```suggestion
2. Add the following code after the `producer_task` in `kafka_example_dag_1`.
```

    ```python
    consumer_logger = logging.getLogger("airflow")
    def consumer_function(message, prefix=None):
        try:
            key = json.loads(message.key())
            value = json.loads(message.value())
            consumer_logger.info(f"{prefix} {message.topic()} @ {message.offset()}; {key} : {value}")
            return
        except:
            return

    consumer_task = ConsumeFromTopicOperator(
        task_id=f"consume_from_{my_topic}",
        topics=[my_topic],
        apply_function=functools.partial(consumer_function, prefix="consumed:::"),
        consumer_config={
            **connection_config
            "group.id": "consume",
            "enable.auto.commit": False,
            "auto.offset.reset": "beginning",
        },
        max_messages=30,
        max_batch_size=10,
    )

    producer_task >> consumer_task

    ```

>     In this code snippet first a function is defined that reads from the topic defined as `my_topic` and prints the messages consumed to the Airflow task log. The ConsumeFromTopicOperator uses the same `connection_config` as the `producer_task` with added configurations specific to Kafka Consumers. You can read more about consumer configuration in Kafka in the official [Kafka documentation](https://kafka.apache.org/documentation/#consumerconfigs).

Reviewer 2: ```suggestion
   The `consumer_task` includes a function that reads from `my_topic` and prints the messages it consumes to the Airflow task log. The `consumer_task` uses the same `connection_config` as the `producer_task` with added configurations specific to Kafka Consumers. You can read more about consumer configuration in Kafka in the [Kafka documentation](https://kafka.apache.org/documentation/#consumerconfigs).
```

3. Run your DAG.

4. View the consumed messages in your Airflow task logs.

    ![Consumer log](/img/guides/kafka-consumer-logs.png)

> :::info

Reviewer 2: ```suggestion
:::tip
```

> A very common pattern is to directly connect an Amazon S3 bucket to your Kafka topic as a consumer. The ConsumeFromTopicOperator is helpful if you want to use Airflow features to schedule the consuming task. Instead of writing the messages retrieved to the Airflow logs, you can write them to S3 using the [S3CreateObjectOperator](https://registry.astronomer.io/providers/amazon/modules/s3createobjectoperator).

Reviewer 2: ```suggestion
A common use case is to directly connect an Amazon S3 bucket to your Kafka topic as a consumer. The ConsumeFromTopicOperator is helpful if you want to use Airflow to schedule the consuming task. Instead of writing the messages retrieved to the Airflow logs, you can write them to S3 using the [S3CreateObjectOperator](https://registry.astronomer.io/providers/amazon/modules/s3createobjectoperator).
```

:::

## Step 4: Listen for a message in the stream

A common use case is to run a downstream task when a specific message appears in your Kafka topic. The AwaitKafkaMessageOperator is a deferrable operator that will listen to your Kafka topic for a message that fulfills a specific criteria.

> :::info
> 
> A deferrable operator is a sensor that will go into a deferred state in between checking for its condition in the target system. While in the deferred state the operator does not take up a worker slot, offering a significant efficiency improvement. Using a deferrable operator necessitates the presence of the Triggerer component. Learn more about [Deferrable operators]((https://docs.astronomer.io/learn/deferrable-operators)).
> 
> :::

Reviewer 2: ```suggestion
A deferrable operator is a sensor that will go into a deferred state in between checking for its condition in the target system. While in the deferred state the operator does not take up a worker slot, offering a significant efficiency improvement. See [Deferrable operators]((https://docs.astronomer.io/learn/deferrable-operators)).
```

1. In `kafka_example_dag_1`, add the following import statement:

    ```python
    from airflow_provider_kafka.operators.await_message import AwaitKafkaMessageOperator
    ```

2. Copy and past the following code at the end of your DAG:

    ```python
    def await_function(message):
        if isinstance(json.loads(message.value()), int):
            if json.loads(message.value()) % 5 == 0:
                return f" Got the following message: {json.loads(message.value())}"

    await_message = AwaitKafkaMessageOperator(
        task_id=f"awaiting_message_in_{my_topic}",
        topics=[my_topic],
        # the apply function needs to be passed with its location for the triggerer
        apply_function="provider_example_code.await_function", 
        kafka_config={
            **connection_config
            "group.id": "awaiting_message",
            "enable.auto.commit": False,
            "auto.offset.reset": "beginning",
        },
        xcom_push_key="retrieved_message",
    )

    consumer_task >> await_message
    ```

>     In this code snippet an `await_function` is defined which will parse each message in the Kafka topic and return the message if the value is an integer divisible by 5. The AwaitKafkaMessageOperator using runs this function over messages polled from the Kafka topic. If no matching message is found it goes into a deferred state until it tries again.  

Reviewer 2: ```suggestion
    This code snippet includes an `await_function` which will parse each message in the Kafka topic and return the message if the value is an integer divisible by 5. The AwaitKafkaMessageOperator using runs this function over messages polled from the Kafka topic. If no matching message is found, it continues to poll in a deferred state until it finds one.
```

Reviewer 2: Is the last sentence right? I thought that it's always trying even if it's in a deferred state.

3. Run the DAG. Notice how the task instance of the `await_message` task goes into a deferred state (purple square).

    ![Kafka deferred state](/img/guides/kafka-deferred-state.png)

4. (Optional) Add a downstream task to the `await_message` task, which only runs once `await_message` has completed successfully.

## How it works

The `airflow-kafka-provider` contains three hooks:

- `KafkaAdminClientHook`
- `KafkaConsumerHook`
- `KafkaProducerHook`

> as well as three operators and one trigger:

Reviewer 2: ```suggestion
It uses three operators and one trigger:
```

- `ProduceToTopicOperator`
- `ConsumeFromTopicOperator`
- `AwaitKafkaMessageOperator`
- `AwaitMessageTrigger`

> in this section we will provide more detailled information on the parameters of the operators. For more information on the other modules in this provider see the [`airflow-provider-kafka`](https://github.com/astronomer/airflow-provider-kafka) source code.

Reviewer 2: ```suggestion
The following section provides more detailed information on the parameters of each operator. For more information on the other modules in this provider see the [`airflow-provider-kafka`](https://github.com/astronomer/airflow-provider-kafka) source code.
```

### ProduceToTopicOperator

The ProduceToTopicOperator can be used create a Kafka producer to produce messages to a Kafka topic. You can define the following parameters:

> - `topic`: the Kafka topic you want to produce to.

Reviewer 2: Capitalize the first letter of all these descriptions.
